Document "no duplicates" enforcement.

jsha opened the pull request:
Also, add belt-and-suspenders checking for serials already existing at
issuance time.

cpu commented on ca/ca.go:
```diff
@@ -571,17 +573,44 @@ func (ca *CertificateAuthorityImpl) IssuePrecertificate(ctx context.Context, iss
+		return emptyCert, err
+	} else if !db.IsNoRows(err) {
+		return emptyCert, fmt.Errorf("error checking for duplicate issuance of %s: %s", serialHex, err)
+	}
```
It looks like `sa.GetCertificate` uses `db.IsNoRows` internally to transform a no-rows `db.ErrDatabaseOp` into a `berrors.NotFoundError` that gets returned indicating the serial wasn't found:

https://github.com/letsencrypt/boulder/blob/e9e15c9a832282c068237a3cdb31c1d91e0de986/sa/sa.go#L361-L364

The `berrors.NotFoundError` instance will always get `false` from `db.IsNoRows` when your code checks it here and that ends up breaking CI when a no-results response from `ca.sa.GetCertificate` is treated as an unexpected error checking for the duplicate:
```
I012549 boulder-ra n83H5Aw [AUDIT] Certificate request - error JSON={"ID":"Z1UKmuLppZ5PLNB3FvbH3C2Ch2NCIaSP0xK12v38yR4","Requester":2,"OrderID":1,"VerifiedFields":["subject.commonName","subjectAltName"],"Names":["rand.3d2a8b8c.xyz"],"NotBefore":"0001-01-01T00:00:00Z","NotAfter":"0001-01-01T00:00:00Z","RequestTime":"2019-06-10T01:24:34Z","ResponseTime":"2019-06-10T01:24:34Z","Error":"issuing certificate for precertificate: rpc error: code = Unknown desc = error checking for duplicate issuance of ff6b088b57833e5ac2597d87f2365c8f85fb: certificate with serial \"ff6b088b57833e5ac2597d87f2365c8f85fb\" not found","Authorizations":{"rand.3d2a8b8c.xyz":{"ID":"1","ChallengeType":"dns-01"}}}
```

I think you want this instead:
```
} else if !berrors.Is(err, berrors.NotFound) {
...
```

jsha replied:
I tried this and got:
```
    ca_test.go:893: Failed to issue cert from precert: error checking for duplicate issuance of 11d8536fc4c4b8936a42e9cbb0c7102a1113: failed to : sql: no rows in result set
```

I think I might need to plumb through a little more. It looks like this is one of the new wrapped database errors. Will give this a shot tomorrow. Pushed my work-in-progress for now (will fail tests).

cpu commented on ca/ca_test.go:
```diff
@@ -904,7 +910,49 @@ func TestIssueCertificateForPrecertificate(t *testing.T) {
+	})
+	if err == nil {
+		t.Fatal("Expected error issuing duplicate serial but got none.")
+	}
```
Does the err generated in this case have any distinguishing features to test for here? e.g. type of error or message?

jsha replied:
Good call! Added a string test. Thanks.

rolandshoemaker commented on ca/ca.go:
```diff
@@ -571,17 +572,43 @@ func (ca *CertificateAuthorityImpl) IssuePrecertificate(ctx context.Context, iss
 	}
+
+	serialHex := core.SerialToString(precert.SerialNumber)
+	if _, err = ca.sa.GetCertificate(ctx, serialHex); err == nil {
```
Is `err == nil` not the successful case? Seems like this should be `err != nil && db.IsDuplicate(err)`.

rolandshoemaker commented:
Test failures appear to be a bug with `IsNoRows` not properly detecting a sql.ErrNoRows (presumably because it's not being wrapped correctly somewhere I think).

rolandshoemaker commented:
Oh huh, `mockSA` the CA test always returns a `db.ErrDatabaseOp{Err: sql.ErrNoRows}` for GetCertificate, which isn't realistic, it should be returning a `berrors.NotFoundError(...)`, this causes the breakage.

jsha commented:
Fixed, and also added some more test fixes and better test coverage. Big thanks to @cpu for recommending that I do better checking on the type of error, since that caught a bug in the test! I was getting an unrelated error because I wasn't providing enough SCTs.

rolandshoemaker commented on ca/ca.go:
```diff
@@ -571,17 +572,43 @@ func (ca *CertificateAuthorityImpl) IssuePrecertificate(ctx context.Context, iss
 	}
+
+	serialHex := core.SerialToString(precert.SerialNumber)
+	if _, err = ca.sa.GetCertificate(ctx, serialHex); err == nil {
```
We also have `SerialExists` now, which doesn't require actually getting the cert, not really a blocker though.

rolandshoemaker replied:
Although I guess we add the serial when we add the precert, so I guess this doesn't do what we'd want... oh well.
